Replace TensorMeta with FakeTensor

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #78836

Alrighty this one is a doozy so listen up.

- FakeTensor by eellison is a better version of TensorMeta, with the primary difference that it makes use of Meta definitions in the dispatcher rather than directly reading out the meta field on the function callable. This is an important improvement because it means that we can use FakeTensor with conventional torch API calls whereas we could only use TensorMeta with refs. A number of previously failing python_ref_consistency tests are now passing, and post this change I recommend using torch API in refs as much as possible (E.g. when you are not making use of new functionality that only occurs in refs). There is no downside and the upside is that it is possible to get fused eager kernels without being forced to decomposed
- To avoid having to wobble all of the Prim metas, I keep a function named TensorMeta that has the old API for constructing fake tensors. TBH all of these should be using methods like new_empty instead
- I need an API for excluding the autograd dispatch key from TLS; the standard `no_grad` doesn't cut it. I exposed the low level TLS functions but on advice by alband we should probably expose the higher level API.
- How prims get registered to dispatcher has changed: previously their implementations were registered as CompositeImplicitAutograd but that prevents prims from getting traced which is no good. So now they are CompositeExplicitAutograd. A consequence of this is that AD on prims stops working but this makes sense since previously we were relying on their aten implementations to implicitly provide AD, which is not long term how it should be implemented. This also means prims are going to show up in traces so backends be ready (I wonder if we need a BC lower back to backends pass). Because backwards no longer works I have to disable it in the OpInfos.
- ProxyTensor is annoyingly recording overload packets to FX IR rather than overloads, which means I can't find the `impl_nvfuser` properties. But since prims are guaranteed to have only one overload I just spray the properties on both the overload and the overload packet
- The PrimContext tracer is now completely subsumed by the ProxyTensor tracer. Well actually I made things worse since ProxyTensor isn't using fake tensors for metadata (but this is the plan)
- Since PrimContext is dead I also had to update execute to take a GraphModule directly instead. But wait there's more: for some reason the old algorithm for constructing nvFuser graphs doesn't work with ProxyTensor's FX graphs, so I rewrote it from scratch to use an FX interpreter rather than mutating the FX graph. Newer implementation is shorter and more correct.
- I simplified TensorLikeType to be literally Tensor (actually we could have done this before but forgot). However this triggered some mypy errors in `atleast_1d` and friends. I fixed it by writing a more verbose implementation which I think is more accurate. In general, however, I feel that it is hard to replicate overload behavior purely in Python
- I relaxed FakeTensor to work with meta tensor inputs cuz this codepath gets triggered in PrimTorch tests
- We are starting to hit some import cycles so there are some shenanigans to keep Python mypy happy

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Replace TensorMeta with FakeTensor

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

albanD commented on torch/csrc/utils/python_dispatch.cpp:
```diff
@@ -226,6 +226,13 @@ void initDispatchBindings(PyObject* module) {
     return states;
   });
 
+  m.def("_dispatch_tls_set_dispatch_key_excluded", [](const char* dispatch_key, bool desired_state) {
```
This should "exclude everything from top to dispatch_key" right?

ezyang replied:
Well the c10 api it binds just pokes out only one dispatch key...

albanD replied:
Yes, but the way you use it, I think you want the exclude all the way to me semantic.
Just like autograd does  AutoDispatchBelowADInplaceOrView guard; You actually want to skip all the top keys as well.

albanD commented on torch/csrc/utils/python_dispatch.cpp:
```diff
@@ -226,6 +226,13 @@ void initDispatchBindings(PyObject* module) {
     return states;
   });
 
+  m.def("_dispatch_tls_set_dispatch_key_excluded", [](const char* dispatch_key, bool desired_state) {
```
I would argue that binding `ExcludeDispatchKeyGuard` will make the API harder to use wrong?
Or do we have a concern with making RAII in python due to potential cycle/gc?

ezyang replied:
It's just more annoying to bind C++ RAII stuff. Actually as you can see from the usage later this isn't good anyway because we have a bunch of constants that are used in conjunction with the guards to actually grab the right dispatch keys. Some API design will be necessary.

albanD replied:
Well binding an RAII is pretty simple: https://github.com/pytorch/pytorch/blob/83d40a4dba54541340462301fc86118167608cb2/torch/csrc/autograd/init.cpp#L397-L398

It is used at https://github.com/pytorch/pytorch/blob/83d40a4dba54541340462301fc86118167608cb2/torch/overrides.py#L1935-L1940

ezyang committed:
Update on "Replace TensorMeta with FakeTensor"


Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

albanD reviewed (approved):
Stamp to unblock but I'm not convinced that the private binding is doing what you want.

ezyang committed:
Update on "Replace TensorMeta with FakeTensor"


Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Replace TensorMeta with FakeTensor"


Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Replace TensorMeta with FakeTensor"


Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Replace TensorMeta with FakeTensor"


Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Replace TensorMeta with FakeTensor"


Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Replace TensorMeta with FakeTensor"


Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Replace TensorMeta with FakeTensor"


Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Replace TensorMeta with FakeTensor"


Alrighty this one is a doozy so listen up.

- FakeTensor by eellison is a better version of TensorMeta, with the primary difference that it makes use of Meta definitions in the dispatcher rather than directly reading out the meta field on the function callable. This is an important improvement because it means that we can use FakeTensor with conventional torch API calls whereas we could only use TensorMeta with refs. A number of previously failing python_ref_consistency tests are now passing, and post this change I recommend using torch API in refs as much as possible (E.g. when you are not making use of new functionality that only occurs in refs). There is no downside and the upside is that it is possible to get fused eager kernels without being forced to decomposed
- To avoid having to wobble all of the Prim metas, I keep a function named TensorMeta that has the old API for constructing fake tensors. TBH all of these should be using methods like new_empty instead
- I need an API for excluding the autograd dispatch key from TLS; the standard `no_grad` doesn't cut it. I exposed the low level TLS functions but on advice by alband we should probably expose the higher level API.
- How prims get registered to dispatcher has changed: previously their implementations were registered as CompositeImplicitAutograd but that prevents prims from getting traced which is no good. So now they are CompositeExplicitAutograd. A consequence of this is that AD on prims stops working but this makes sense since previously we were relying on their aten implementations to implicitly provide AD, which is not long term how it should be implemented. This also means prims are going to show up in traces so backends be ready (I wonder if we need a BC lower back to backends pass). Because backwards no longer works I have to disable it in the OpInfos.
- ProxyTensor is annoyingly recording overload packets to FX IR rather than overloads, which means I can't find the `impl_nvfuser` properties. But since prims are guaranteed to have only one overload I just spray the properties on both the overload and the overload packet
- The PrimContext tracer is now completely subsumed by the ProxyTensor tracer. Well actually I made things worse since ProxyTensor isn't using fake tensors for metadata (but this is the plan)
- Since PrimContext is dead I also had to update execute to take a GraphModule directly instead. But wait there's more: for some reason the old algorithm for constructing nvFuser graphs doesn't work with ProxyTensor's FX graphs, so I rewrote it from scratch to use an FX interpreter rather than mutating the FX graph. Newer implementation is shorter and more correct.
- I simplified TensorLikeType to be literally Tensor (actually we could have done this before but forgot). However this triggered some mypy errors in `atleast_1d` and friends. I fixed it by writing a more verbose implementation which I think is more accurate. In general, however, I feel that it is hard to replicate overload behavior purely in Python
- I relaxed FakeTensor to work with meta tensor inputs cuz this codepath gets triggered in PrimTorch tests
- We are starting to hit some import cycles so there are some shenanigans to keep Python mypy happy

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

mruberry commented on torch/_prims/__init__.py:
```diff
@@ -286,27 +289,29 @@ class RETURN_TYPE(Enum):
     return wrapper
 
 
+@contextlib.contextmanager
```
Comment here about what this function does and how it's intended to be used

mruberry commented on torch/_prims/__init__.py:
```diff
@@ -330,16 +335,33 @@ def _prim_impl(*args, **kwargs):
         meta(*args, **kwargs)
         return impl_aten(*args, **kwargs)
 
+    class BackwardsNotSupported(torch.autograd.Function):
```
Comment for this class

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -19378,6 +19378,12 @@ def _inherit_constructor_args(name, op, inherited, overrides):
     kwargs.update(common_kwargs)
     kwargs.update(overrides)
 
+    kwargs['supports_autograd'] = False
```
TODO comment here

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -20059,10 +20056,6 @@ def __init__(
     PythonRefInfo(
         "_refs.t",
         torch_opinfo_name="t",
-        decorators=(
```
Nice

mruberry reviewed (approved):
Ship it!

Some inline comment suggestions but we needn't wait for them (more comments in general, please!).

General note: seems like "FakeTensor" should just replace "MetaTensor" and be called "MetaTensor"?

ezyang commented:
@pytorchbot merge
